rust/src/ja4.rs

```diff
@@ -104,104 +63,64 @@ impl JA4 {
         }
     }
 
-    pub fn new() -> Self {
-        Self {
-            tls_version: None,
-            ciphersuites: Vec::with_capacity(20),
-            extensions: Vec::with_capacity(20),
-            signature_algorithms: Vec::with_capacity(20),
-            domain: false,
-            alpn: ['0', '0'],
-            quic: false,
-            nof_exts: 0,
-        }
-    }
-
-    pub fn set_quic(&mut self) {
-        self.quic = true;
-    }
-
-    pub fn set_tls_version(&mut self, version: TlsVersion) {
-        if JA4::is_grease(u16::from(version)) {
-            return;
-        }
-        // Track maximum of seen TLS versions
-        match self.tls_version {
-            None => {
-                self.tls_version = Some(version);
-            }
-            Some(cur_version) => {
-                if u16::from(version) > u16::from(cur_version) {
-                    self.tls_version = Some(version);
+    fn format_alpn(alpn: Option<&Vec<u8>>) -> [char; 2] {
+        let mut ret = ['0', '0'];
+
+        if let Some(alpn) = alpn {
+            if !alpn.is_empty() {
+                // If the first ALPN value is only a single character, then that character is treated as both the first and last character.
+                if alpn.len() == 2 {
+                    // GREASE values are 2 bytes, so this could be one -- check
+                    let v: u16 = ((alpn[0] as u16) << 8) | alpn[alpn.len() - 1] as u16;
+                    if HandshakeParams::is_grease(v) {
+                        return ret;
+                    }
                 }
-            }
-        }
-    }
-
-    pub fn set_alpn(&mut self, alpn: &[u8]) {
-        if !alpn.is_empty() {
-            // If the first ALPN value is only a single character, then that character is treated as both the first and last character.
-            if alpn.len() == 2 {
-                // GREASE values are 2 bytes, so this could be one -- check
-                let v: u16 = ((alpn[0] as u16) << 8) | alpn[alpn.len() - 1] as u16;
-                if JA4::is_grease(v) {
-                    return;
+                if !alpn[0].is_ascii_alphanumeric() || !alpn[alpn.len() - 1].is_ascii_alphanumeric()
+                {
+                    // If the first or last byte of the first ALPN is non-alphanumeric (meaning not 0x30-0x39, 0x41-0x5A, or 0x61-0x7A), then we print the first and last characters of the hex representation of the first ALPN instead.
```

Comment:
is this rustfmt compliant?

Reply:
This functionality and format was already present in the codebase. I just moved it to a function as it helped with the code I was writing. It was already rustfmt compliant :)

Reply:
Ok cool. Just surprised about the long comment line :)